archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java

```diff
@@ -1340,48 +1225,37 @@ private void completeInterfacesFrom(ImportContext context) {
 
     void completeEnclosingClassFrom(ImportContext context) {
         enclosingClass = context.createEnclosingClass(this);
+        completionProcess.markEnclosingClassComplete();
     }
 
     void completeTypeParametersFrom(ImportContext context) {
         typeParameters = context.createTypeParameters(this);
+        completionProcess.markTypeParametersComplete();
     }
 
     void completeGenericSuperclassFrom(ImportContext context) {
         Optional<JavaType> genericSuperclass = context.createGenericSuperclass(this);
         if (genericSuperclass.isPresent()) {
             superclass = superclass.withGenericType(genericSuperclass.get());
         }
+        completionProcess.markGenericSuperclassComplete();
     }
 
     void completeMembers(final ImportContext context) {
-        fields = context.createFields(this);
-        methods = context.createMethods(this);
-        constructors = context.createConstructors(this);
-        staticInitializer = context.createStaticInitializer(this);
-        codeUnits = ImmutableSet.<JavaCodeUnit>builder()
-                .addAll(methods).addAll(constructors).addAll(staticInitializer.asSet())
-                .build();
-        members = ImmutableSet.<JavaMember>builder()
-                .addAll(fields)
-                .addAll(methods)
-                .addAll(constructors)
-                .build();
+        members = JavaClassMembers.create(this, context);
```

Comment:
That's much better than the `Method may not be called before construction of hierarchy is complete` exceptions, thank you!